src/defs.hpp

```diff
@@ -56,21 +60,211 @@ struct LogicalLocation { // aggregate and POD type
+  // volume
+  bool IsNeighbor(const LogicalLocation &in) const {
+    if (in.level() < level()) return in.IsNeighbor(*this);
+    if (Contains(in)) return false; // You share a volume
+    // Only need to consider case where other block is equally or more refined than you
+    auto offset = 1 << (in.level() - level());
+    const auto shifted_lx1 = lx1_ << (in.level() - level());
+    const auto shifted_lx2 = lx2_ << (in.level() - level());
+    const auto shifted_lx3 = lx3_ << (in.level() - level());
+    const bool bx1 =
+        (in.lx1() >= (shifted_lx1 - 1)) && (in.lx1() <= (shifted_lx1 + offset));
+    const bool bx2 =
+        (in.lx2() >= (shifted_lx2 - 1)) && (in.lx2() <= (shifted_lx2 + offset));
+    const bool bx3 =
+        (in.lx3() >= (shifted_lx3 - 1)) && (in.lx3() <= (shifted_lx3 + offset));
+    return bx1 && bx2 && bx3;
+  }
+
+  LogicalLocation GetSameLevelNeighbor(int ox1, int ox2, int ox3) const {
+    return LogicalLocation(level_, lx1_ + ox1, lx2_ + ox2, lx3_ + ox3);
+  }
+
+  LogicalLocation GetParent() const {
+    if (level_ == 0) return *this;
+    return LogicalLocation(level_ - 1, lx1_ >> 1, lx2_ >> 1, lx3_ >> 1);
+  }
+
+  std::vector<LogicalLocation> GetDaughters() const {
+    std::vector<LogicalLocation> daughters;
+    for (int i : {0, 1}) {
+      for (int j : {0, 1}) {
+        for (int k : {0, 1}) {
+          daughters.push_back(GetDaughter(i, j, k));
+        }
+      }
+    }
+    return daughters;
+  }
+
+  LogicalLocation GetDaughter(int ox1, int ox2, int ox3) const {
+    return LogicalLocation(level_ + 1, (lx1_ << 1) + ox1, (lx2_ << 1) + ox2,
+                           (lx3_ << 1) + ox3);
+  }
+
+  std::set<LogicalLocation>
+  GetPossibleBlocksSurroundingTopologicalElement(int ox1, int ox2, int ox3) const {
+    std::vector<LogicalLocation> locs;
+
+    const auto irange =
+        (std::abs(ox1) == 1) ? std::vector<int>{0, ox1} : std::vector<int>{0};
+    const auto jrange =
+        (std::abs(ox2) == 1) ? std::vector<int>{0, ox2} : std::vector<int>{0};
+    const auto krange =
+        (std::abs(ox3) == 1) ? std::vector<int>{0, ox3} : std::vector<int>{0};
+
+    auto AddNeighbors = [&](const LogicalLocation &loc) {
+      int n_cells_level = std::pow(2, loc.level());
+      for (int i : irange) {
+        for (int j : jrange) {
+          for (int k : krange) {
+            const auto lx1 = loc.lx1() + i;
+            const auto lx2 = loc.lx2() + j;
+            const auto lx3 = loc.lx3() + k;
+            // TODO(LFR): Deal with periodic boundaries, maybe a little complicated
+            // because of root grid stuff
+            if (0 <= lx1 && lx1 < n_cells_level && 0 <= lx2 && lx2 < n_cells_level &&
+                0 <= lx3 && lx3 < n_cells_level) {
+              locs.emplace_back(loc.level(), lx1, lx2, lx3);
+              auto parent = locs.back().GetParent();
+              if (IsNeighbor(parent)) locs.push_back(parent);
+            }
+          }
+        }
+      }
+    };
```

Comment:
Should this definitely be a lambda and not its own function?

Reply:
Well, it relies on capturing `irange`, `krange`, and `jrange` which depend on the offsets (as well as `loc`). I guess it would make sense to call it `GetSameLevelOrCoarserTopologicalElementNeighbors` and just pass `ox1`, `ox2` and `ox3` to it and calculate the ranges internally.

Reply:
I'm fine with it living where it currently does. Just wondered if it might belong elsewhere. Either way is fine.

Reply:
I think I will leave it here for now.